src/ic-cdk-macros/tests/imports.rs

```diff
@@ -0,0 +1,14 @@
+// https://github.com/dtolnay/trybuild/issues/95
+use ic_cdk_macros::import;
+
+#[import(canister_id = "aaaaa-aa", candid_path = "tests/ic.did")]
+mod mgmt {
```

Comment:
What exactly is generated here? Are you trying to add functions to the existing module, or check for subtyping of the given interface?

Can you add a test for recursive data structure? For example, what does `type List = opt record { int; List }` translates to Rust?

Reply:
> Adding members to the module overrides existing ones with custom definitions

I think we are coming to the same conclusion: it's better to let users write down their type definitions, and the proc macro will just check for subtype. If the user didn't write a particular method or definition in the module, it gets ignored, not imported. The type name, derive traits, `Box` vs `Rc` vs lifetime are all hard to control.

> there is a function to validate a subtype, but it requires a TypeEnv forest and other things that are convenient to fetch during parsing but not afterwards

Not sure what you mean here. `add_decl` is basically your version of `TypeEnv`, why not use `TypeEnv` directly?

Reply:
If we only go off of explicit declarations then that's vastly more boilerplate that the macro is capable of generating. It already works basically perfectly as-is, with auto-imported but overridable defaults. As long as there is a default mapping for every possible candid file, and the codegen can be treated as set-it-and-forget-it, that's much less friction for the average user - that was the primary reason I started this, so I wouldn't be stuck writing boilerplate. Anything that might be sub-optimal in a particular scenario, like Nat when the user would be happy writing u128, can just be overridden, and if they have to define that themselves to begin with then AFAICT that's strictly less functionality.
> why not use `TypeEnv` directly?
Mostly because I hadn't thought of it, and I had only been looking on the methods of TypeEnv for its functionality instead of noticing the free functions in the surrounding module. I now think type-checking will be a lot easier than I thought it would a bit ago.

Reply:
Actually, on second thought, I think it *is* still going to be hard, because the Candid type of a Rust type is only accessible at runtime as long as it's a trait function. The maximum it would be possible to do is crash on init if the types are wrong, and even that would require changing how `#[init]` works.

Reply:
>  I wouldn't be stuck writing boilerplate

If that's the goal, generating an initial binding from `didc` would be much easier. The user can then modify the file as they wish. Proc macro is hard to debug and understand, and the generated code is not visible to the user. Only use proc macro when the generated code is indeed transparent to the user. But in our case, it is not.

> because the Candid type of a Rust type is only accessible at runtime as long as it's a trait function.

Combined with dfx, we can export the candid type and check for subtype. See #28

Reply:
See other discussion for macros, but re: dfx exporting, if an external command can generate a DID file from Rust code that's perfect, but then the actual type checking seems like it'd belong in that command or some auxiliary command, as the macro never actually operates on the code it's touching, only operating on files generated on that code by an external tool. And if the macro were to call that tool, then building the crate once would require building it twice.